### PR TITLE
Fire - Earth DREAM TEAM (Mona, Kalki, Emily, Anya)

### PR DESCRIPTION
**Pull request by anyatokar**
# Assignment Submission: bEtsy

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions. These should be answered by all members of your team, not by a single teammate.

## Reflection

Prompt | Response
--- | ---
Each team member: what is one thing you were primarily responsible for that you're proud of? | Mona: Leave a review for a product. Anya: db seeds/validations were primarily me but I am more proud of my model method and view page contributions. Kalki: User and Cart functionalities. Emily: HTML/CSS
Each team member: what is one thing you were primarily responsible for that you would like targeted feedback on? | Mona: review controller test. Anya: would like feedback on seeds and edge case testing. Kalki: Methods used for order items and cart, if there are simpler and more efficient ways to do things. Emily: Primarily responsible for html/css. Any tips on how to refactor css and lowering the footer.
How did your team break up the work to be done? | We posted requirements to Trello. We each chose pieces of the project we were most excited/cared the most about and worked on those.
How did your team utilize git to collaborate? | We submitted pull requests when making big updates.
What did your group do to try to keep your code DRY while many people collaborated on it? | We reviewed each other’s code.
What was a technical challenge that you faced as a group? | A technical challenge was figuring out how to code the cart.
What was a team/personal challenge that you faced as a group? | A personal team challenge was having to do everything over zoom instead of in person. We ran into issues merging CSS at the end.
What was your application's ERD? (upload this to Google Drive, change the share settings to viewable by everyone with a link, and include a link) | https://lucid.app/lucidchart/b3a97af0-1592-4364-8433-564986784f9e/edit?shared=true&page=0_0#?folder_id=home&browser=iconWhat is your Trello URL? | https://trello.com/b/MN4fqgSZ/dream-team
What is the Heroku URL of your deployed application? | https://dreamsy.herokuapp.com/ (this version doesn't have the css yet)

### CODE REVIEW

File: db/order_items_seeds.csv [hunk 1]
```diff
@@ -0,0 +1,21 @@
+name,credit_card_name,quantity,shipped
+Travel the world,Sammy D'Amore,3,true
+Walk on the moon,Fred Runolfsson,2,true
+Find your soulmate,Sammy D'Amore,1,true
+Buy an island,Jesse Luettgen,3,true
+Bottle of joy,Amb. Glenda Legros,1,false
+Bottle of love,Elmer Shields,1,true
```
Comment (beccaelenzil) [inline db/order_items_seeds.csv:7]: There's no 'bottle of love' product, so I removed these from the seeds in a few places so that it would fully seed.

File: test/controllers/reviews_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  let(:review_data) {
+    {
+        review: {
+            rating: 3,
+            review_content: "A good product!",
+            name: "test name",
+        }
+    }
+  }
+  describe "users(Guest)" do
```
Comment (beccaelenzil) [inline test/controllers/reviews_controller_test.rb:14]: Mona requested feedback: Good work dividing the tests into logged in users and guests. Also good use of fixtures.

File: test/controllers/reviews_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        expect(saved_review.name).must_equal review_data[:review][:name]
+      end
+
+      it "does not create a review if data is missing" do
+        review_data[:review][:rating] = nil
+        expect {
+          post product_reviews_path(product_id), params: review_data
+        }.must_differ "Review.count", 0
+        must_respond_with :redirect
+      end
```
Comment (beccaelenzil) [inline test/controllers/reviews_controller_test.rb:34]: An additional edge case to consider is not being able to create a view for a retired product.

File: app/models/order.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  def update_order_status
+    if self.status == 'pending'
+      self.update_attribute(:status, 'paid')
+    end
+  end
+
+  def update_stock
+    self.order_items.each do |item|
+      item.product.stock -= item.quantity
+      product.stock.update_attribute(:stock, item.product.stock)
```
Comment (beccaelenzil) [inline app/models/order.rb:54]: It looks like this method is never called. That is probably why stock is not updating

File: app/models/order.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  validates :city, presence: { message: "City is required" }, :on => :update
+  validates :state, presence: { message: "State is required" }, :on => :update
+  validates :zip_code, length: { is: 5, message: "Enter valid 5-digit zip code" }, :on => :update# doesn't check for digits only
+
+
+  def add_product(product, quantity)
+    current_item = OrderItem.find_by(product_id: product.id, order_id: self.id )
+
+    if current_item
+      current_item.quantity = quantity
```
Comment (beccaelenzil) [inline app/models/order.rb:20]: Quantity is not updated. This looks like a small bug.

```suggestion
      current_item.quantity += quantity
```

File: test/models/order_test.rb [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+  describe "total" do
+    it "calculates order total" do
+      order = orders(:order1)
+      result = order.total
+      expect(result).must_equal 8
+    end
+
+    # it "when order is nil or empty" do # TODO not sure how to write this
+    #   order = nil
```
Comment (beccaelenzil) [inline test/models/order_test.rb:294]: First create an Order with no Order items.

File: test/models/order_test.rb [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+      product3 = products(:product3)
+      order5 = orders(:order5) # order5 already has 3 product3s in it
+      quantity = 3
+
+      expect(OrderItem.count).must_equal 5
+      order5.add_product(product3, quantity)
+      expect(OrderItem.count).must_equal 5
+    end
+
+    it "cant add item to cart if its retired" do # TODO perhaps move this to order_items controller
```
Comment (beccaelenzil) [inline test/models/order_test.rb:258]: This is a great test to have. It seems that you ran out of time to complete this?

File: test/models/user_test.rb [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+      result1 = user1.total_rev_filtered("paid")
+      result2 = user1.total_rev_filtered("pending")
+      expect(result1).must_equal 10
+      expect(result2).must_equal 2
+    end
+
+    # TODO perhaps test for an unexpected value for status
+  end
+
+  describe "total_orders_by_status" do
```
Comment (beccaelenzil) [inline test/models/user_test.rb:99]: Consider the edge case for user that has no orders.

File: test/models/user_test.rb [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    it "if status is other expected value such as pending and paid" do
+      user1 = users(:user1)
+      result1 = user1.total_orders_by_status("paid")
+      result2 = user1.total_orders_by_status("pending")
+      expect(result1).must_equal 2
+      expect(result2).must_equal 1
+    end
+  end
+
+  describe "sorted_products" do
```
Comment (beccaelenzil) [inline test/models/user_test.rb:116]: Consider the edge case for user that has no products.

File: app/models/order_item.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+  validates :quantity, presence: true
+  validates :quantity, numericality: { only_integer: true, greater_than: 0, message: "The quantity must be a positive value."}
+
+
+  def order_item_subtotal
+    self.quantity * self.product.price
+  end
+
+  def stock_update
```
Comment (beccaelenzil) [inline app/models/order_item.rb:14]: This method isn't used.

File: app/models/order.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  validates :cc_exp_month, numericality: { in: 1..12, message: "Enter valid expiration month (1-12)" }, :on => :update # date validation could be improved with time
+  validates :cc_exp_year, numericality: { in: 2020..2030, message: "Enter valid expiration year (2020-2030)" }, :on => :update
+  validates :email, email: true, :on => :update
+  validates :street_address, presence: { message: "Street address required" }, :on => :update
+  validates :city, presence: { message: "City is required" }, :on => :update
+  validates :state, presence: { message: "State is required" }, :on => :update
+  validates :zip_code, length: { is: 5, message: "Enter valid 5-digit zip code" }, :on => :update# doesn't check for digits only
+
+
+  def add_product(product, quantity)
```
Comment (beccaelenzil) [inline app/models/order.rb:16]: Kalki requested feedback: Good work moving this logic into custom model methods rather than the controller. There are just a couple small bugs outlined below that keep functionality such as updating quantity and stock from working correctly.

**Review by beccaelenzil — COMMENTED** [review]
# bEtsy



## Functional Requirements: Manual Testing

Workflow | yes / no
---   | ---
**Before logging in** |
Browse all products, by category, by merchant | ✔️ Consider a drop down for merchants similar to categories.
Leave a review | ✔️
Verify unable to create a new product | ✔️ I was able to view the new page, but not able to create product. Consider using `require_login` filter to keep guest user from seeing this page.
**After logging in** | |
Create a category | ✔️
Create a product in that category with stock 10 | ✔️
Add the product you created to your cart | ✔️
Add it again (should update quantity) | the quantity was not updated
Verify unable to increase quantity beyond stock | ✔️
Add another merchant's product | ✔️
Check out | ✔️
Check that stock was reduced | stock was not reduced
Change order-item's status on dashboard | ✔️
Verify unable to leave a review for your own product | ✔️
Verify unable to edit another merchant's product by manually editing URL | ✔️
Verify unable to see another merchant's dashboard by manually editing URL | ✔️

## Major Learning Gocode als/Code Review

Criteria | yes / no
---  | ---
90% reported coverage for all controller and model classes using SimpleCov | model test coverage is low
**Routes** |
No un-needed routes generated (check `reviews`) | ✔️
Routes not overly-nested (check products and merchants) | ✔️
Merchant dashboard and cart page use a non-parameterized routes (should pull merchant or cart ID from session) | Instead of using the users show page, we recommend creating a custom route for the merchant dashboard
**Controllers** |
Controller-filter to require login is applied to all merchant-specific actions (update/add item, add category, view merchant dashboard, etc.) - filter method is not duplicated across multiple files | ✔️ yes! but you don't use the require login feature.
Helper methods or filters to find logged-in user, cart, product, etc | ✔️ There are some of these. One that could be useful is `find_product`
No excessive business logic | ✔️
**Business logic that ought to live in the model** |
Add / remove / update product on order | ✔️ See in-line comment for small bug `= => +=`
Checkout -> decrease inventory | ✔️ See in-line comment. It looks like `update_stock` is never called
Merchant's total revenue | ✔️
Find all orders for this merchant (instance method on `Merchant`) | ✔️
**Selected Model Tests** |
Add item to cart:<br>  - Can add a good product<br>  - Can't add a product w/o enough stock<br>  - Can't add a retired product<br>  - Can't add to an order that's not in cart mode<br>  - Logic specific to this implementation | <br>✔️<br>✔️<br>✔️<br>✔️<br>It looks like there are a few started tests that were not completed<br>
Get orders for this merchant:<br>  - Includes all orders from this merchant<br>  - Doesn't include orders from another merchant<br>  - Orders are not included more than once<br>  - Does something reasonable when there are no orders for this merchant | <br>✔️<br>✔️<br>✔️<br>Missing tests for edge cases<br>
**Selected Controller Tests** |
Add item to cart:<br>  - Empty cart (should be created)<br>  - Cart already exists (should add to same order)<br>  - Product already in cart (should update quantity)<br>  - Bad product ID, product is retired, quantity too high, or something like that (error) | <br>✔️<br>This is specifically not tested<br>This feature is not working<br>✔️<br>
Leave a review:<br>  - Works when not logged in<br>  - Works when logged in as someone other than the product's merchant<br>  - Doesn't work if logged in as this product's merchant<br>  - Doesn't work if validations fail | ✔️



## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Overall Feedback

Great work overall! You've built a fully functional web store from top to bottom. This represents a huge amount of work, and you should be proud of yourselves!. 

I am particularly impressed by the way that you wrote clear model methods to encapsulate key functionality. Furthermore you've made good use of controller filters in your application controllers to DRY up your code. The user experience on the site is quite nice, and it's beautifully styled.

I do see some room for improvement around thoroughly testing both controller and model methods. It seems that you may have run out of time as I saw a few tests commented out. I encourage you to review the simplecov code coverage report and consider how the code that is not covered could be tested. There were also a few bugs that affected the functionality of your program (one in the seeds, and a couple others for updating quantity and stock). To work out these bugs, it's best to go through all the requirements from the rubric one last time to insure your code is working as expected.

bEtsy is a huge project on a very short timeline, and this feedback should not at all diminish the magnitude of what you've accomplished. Keep up the hard work!

---

Mona and Kalki I provided in-line feedback for your requested feedback.

Emily requested feedback: The styling is beautiful! I know you learned a lot. I'm not a CSS pro, but I thought I'd share one example of using bootstrap to create nice order forms: https://mdbootstrap.com/snippets/jquery/pjoter-2-0/747937

Anya requested feedback: When testing edge cases it can be very helpful to rely on your simplecov code coverage report. The code in `else` clauses often represent edge cases. What happens when a product or order is nil, for instance. In addition common edge cases are how does a particilar method behave when the array is empty such as a user has no products or orders.







## Only the person who submitted the PR will get an email about this feedback. Please let the rest of your team know about it.
